src/db/spanner/models.rs

```diff
@@ -0,0 +1,1292 @@
+        sqlparams.insert("collectionid".to_string(), collection_id.to_string());
+        sqlparams.insert("bsoid".to_string(), bso.id.to_string());
+        sql.params = Some(sqlparams);
+
+        let results = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        let exists = match results {
+            Ok(results) => results.1.rows.is_some(),
+            // TODO Return the correct error
+            Err(_e) => {
+                return Err(DbErrorKind::CollectionNotFound.into());
+            }
+        };
+        let sql = if exists {
+            let mut sql = ExecuteSqlRequest::default();
+            sql.sql = Some("UPDATE bso SET sortindex=@sortindex, expiry=@expiry, payload=@payload, modified=@modified WHERE user_id = @userid AND collection_id = @collectionid AND id = @bsoid".to_string());
+            let mut sqlparams = HashMap::new();
+            sqlparams.insert("sortindex".to_string(), bso.sortindex.unwrap().to_string());
+            sqlparams.insert(
+                "expiry".to_string(),
+                bso.ttl
+                    .map(|ttl| timestamp + (i64::from(ttl) * 1000))
+                    .unwrap()
+                    .to_string(),
+            );
+            sqlparams.insert("payload".to_string(), bso.payload.unwrap().to_string());
+            sqlparams.insert("modified".to_string(), timestamp.to_string());
+            sqlparams.insert("userid".to_string(), user_id.to_string());
+            sqlparams.insert("collectionid".to_string(), collection_id.to_string());
+            sqlparams.insert("bsoid".to_string(), bso.id.to_string());
+            sql.params = Some(sqlparams);
+            sql
+        } else {
+            let mut sql = ExecuteSqlRequest::default();
+            sql.sql = Some("INSERT INTO bso (userid, collection_id, id, sortindex, payload, modified, expiry) VALUES (@userid, @collection_id, @bsoid, @sortindex, @payload, @modified, @expiry))".to_string());
+            let mut sqlparams = HashMap::new();
+            sqlparams.insert("userid".to_string(), user_id.to_string());
+            sqlparams.insert("collectionid".to_string(), collection_id.to_string());
+            sqlparams.insert("bsoid".to_string(), bso.id.to_string());
+            sqlparams.insert("sortindex".to_string(), bso.sortindex.unwrap().to_string());
+            sqlparams.insert("payload".to_string(), bso.payload.unwrap().to_string());
+            sqlparams.insert("modified".to_string(), timestamp.to_string());
+            sqlparams.insert(
+                "expiry".to_string(),
+                bso.ttl
+                    .map(|ttl| timestamp + (i64::from(ttl) * 1000))
+                    .unwrap()
+                    .to_string(),
+            );
+
+            sql.params = Some(sqlparams);
+            sql
+        };
+
+        let results = spanner
```

Comment:
Is locking a concern here like it was above with inserts?

Reply:
All the writes are guarded by lock_for_write (a Spanner read/write transaction) so we should be "locked here". When a transaction can't succeed we will get an ABORT (on commit, but possibly on other API calls I believe).

We'll have a separate issue to cleanup the error handling. Part of that will be ensuring ABORTs are converted to DbError::ConflictErrors (503 to the client to retry)